Automattic/woocommerce-payments
# Enforce maximum dispute evidence text length

Fixes #290

#### Changes proposed in this Pull Request

Stripe [places character count limits](https://stripe.com/docs/api/disputes/update#update_dispute-evidence) on fields in dispute evidence. This PR enforces those limits when the merchant fills out the dispute evidence form: when the merchant reaches or exceeds the limit (whether by typing or by pasting already copied text), an error notice will appear letting them know they have exceeded the maximum length and the text field won't allow the character(s) that exceed the limit.

<img width="767" alt="Reached maximum character count for evidence" src="https://user-images.githubusercontent.com/16066224/136585255-89cc827b-fc32-49b2-96be-9639ed31b19d.png">

There are two types of Stripe limits enforced by this PR:
* Eight of the fields (e.g. `refund_refusal_explanation` and `uncategorized_text`) have a 20,000 character limit for the field. One of those fields (`activity_access_log`) we have merchants upload a file for instead, so I'm not currently enforcing the character limit there.
* Across all fields, the maximum character count is 150,000.

#### Testing instructions

* In test mode, place an order using a [number that triggers a dispute](https://stripe.com/docs/testing#disputes), e.g. `4000000000002685`.
* Go to **Payments > Disputes**, click on the disputed payment, and choose **Challenge dispute**.
* For each of the disputed fields ("Product description", "Refund policy disclosure", "Refund refusal explanation", "Explanation of duplicate charge", "Cancellation policy disclosure", "Cancellation rebuttal", "Additional details"), entering >= 20,000 characters should trigger the notice. An easy way to test this is with `php -r "echo str_repeat('a', 19995);"` to generate close to 20,000 characters and verify that typing a few more triggers the error notice, or by generating >20,000 characters and verifying that pasting triggers the error notice too.
* To check the across-all-fields limit, paste about 149,900 characters into a non-limited field (e.g. "Customer billing address"). Then, adding a small number of additional characters to any field should trigger the 150,000 character limit (the exact number will vary based on the customer name/email and other pre-filled fields).

**Known issues**

* This PR only handles the client-side validation, but we should probably separately handle this validation server-side too or cleanly handle whatever error Stripe gives back when the limit is exceeded.

-------------------


**Post merge**
